README.md
![build](https://github.com/matyalatte/FF7R-DDS-tools/actions/workflows/build.yml/badge.svg)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

# FF7R-DDS-tools ver0.1.7
Texture mod tools for FF7R<br>
All you need is drop files or folders on batch files.<br>

## Features

- Inject any size DDS and any number of mipmaps.
- Export assets as DDS.
- Remove mipmaps from assets.

## Supported Formats

- DXT1/BC1
- DXT5/BC3
- BC4/ATI1
- BC5/ATI2
- BC6H
- BC7
- B8G8R8A8(sRGB)
- FloatRGBA

## Download
Download `FF7R-DDS-tools*.zip` from [here](https://github.com/matyalatte/FF7R-DDS-tools/releases)

## Basic Usage
1. Drop `.uexp` onto `1_copy_uasset*.bat`.<br>
   The asset will be copied in `./workspace/uasset`.<br>

2. Drop `.dds` onto `2_inject_dds*.bat`.<br>
   A new asset will be generated in `./injected`.<br>

## Advanced Usage
You can inject multiple assets at the same time.<br>
See here for the details.<br>
[Advanced Usage · matyalatte/FF7R-DDS-tools Wiki](https://github.com/matyalatte/FF7R-DDS-tools/wiki/Advanced-Usage)


## Batch files
- `1_copy_uasset*.bat`<br>
    Make or clear `./workspace`.<br>
    Then, copy an asset to workspace.

- `2_inject_dds*.bat`<br>
    Inject dds into the asset copied to workspace.<br>
    A new asset will be generated in `./injected`.

- `_export_as_dds*.bat`<br>
    Export texture assets as dds.<br>

- `_remove_mipmaps*.bat`<br>
    Remove all mipmaps from an asset.<br>
    A new asset will be generated in `./removed`.

- `_parse*.bat`<br>
    Parse files.<br>
    You can check the format with this batch file.

## How to Build
You can build my tool with Github Actions.<br>
See here for the details.<br>
[How to Build with Github Actions · matyalatte/FF7R-mesh-importer Wiki](https://github.com/matyalatte/FF7R-mesh-importer/wiki/How-to-Build-with-Github-Actions)

## FAQ

### Is there any point in removing mipmaps?
Basically no.<br>
But there are minor mipmap problems with model mods.<br>
The removal function will resolve them.

### I got the `UE4 requires BC6H(unsigned)...` warning. What should I do?
There are two types of BC6H: `signed` and `unsigned`.<br>
And you should use the `unsigned` format.<br>
See here for the details.<br>
[How to Inject .HDR textures · matyalatte/FF7R-DDS-tools Wiki](https://github.com/matyalatte/FF7R-DDS-tools/wiki/How-to-Inject-.HDR-textures)

### I got the `Mipmaps should have power of 2 as...` warning. What should I do?
Change its width and height to power of 2.<br>
Or export dds without mipmaps.


changelog.txt
ver 0.1.7
- Add support for .DDS
- Add feature to specify files in txt
- Bug fix (fixed an error when loading some assets)

ver 0.1.6
- Add support for BC7, DXT5/BC3, BC1
- Remove header files

ver 0.1.5
- Ignore sign when importing BC6H
- Add support for B8G8R8A8(sRGB)

ver 0.1.4
 - Add support for BC6H(unsigned)
 - Add support for any size mipmaps (no need to use power of 2 as height and width)

ver 0.1.3
 - Add export function
 - Bugfix

ver 0.1.2
 - Add support for ATI1 and ATI2

ver 0.1.1
 - Allow an folder as input.

ver 0.1.0
 - Initial version

src/main.py
import os, argparse, shutil
from io_util import mkdir, compare
from texture_asset import TextureUasset, get_all_file_path
from dds import DDS
from file_list import get_file_list_from_folder, get_file_list_from_txt, get_file_list_rec

#get arguments
def get_args():
    parser = argparse.ArgumentParser()
    parser.add_argument('file', help='.uasset, .uexp, .ubulk, or a folder')
    parser.add_argument('--save_folder', default='output', type=str, help='save folder')    
    parser.add_argument('--mode', default='parse', type=str, help='valid, parse, copy_uasset, inject, or remove_mipmaps')    
    args = parser.parse_args()
    return args

#parse dds or uasset
def parse(folder, file, save_folder, clear=True):
    file = os.path.join(folder, file)
    if file[-3:] in ['dds', 'DDS']:
        DDS.load(file, verbose=True)
    else:
        TextureUasset(file, verbose=True)

#check if the tool can read and write a file correctly.
def valid(folder, file, save_folder, clear=True):

    #make or clear workspace
    save_folder = 'workspace/valid'
    if clear and os.path.exists(save_folder):
        shutil.rmtree(save_folder)
        print('clear: {}'.format(save_folder))
    mkdir(save_folder)
    print('clear: {}'.format(save_folder))

    src_file = os.path.join(folder, file)
    new_file=os.path.join(save_folder, file)
    
    if file[-3:] in ['dds', 'DDS']:
        #read and write dds
        dds = DDS.load(src_file)
        dds.save(new_file)

        #compare and remove files
        compare(src_file, new_file)
        os.remove(new_file)

    else:
        #read and write uasset
        uasset_name, uexp_name, ubulk_name = get_all_file_path(src_file)
        texture = TextureUasset(file, verbose=True)
        new_uasset_name, new_uexp_name, new_ubulk_name = texture.save(new_file)

        #compare and remove files
        compare(uasset_name, new_uasset_name)
        compare(uexp_name, new_uexp_name)
        os.remove(new_uasset_name)
        os.remove(new_uexp_name)
        if new_ubulk_name is not None:
            compare(ubulk_name, new_ubulk_name)
            os.remove(new_ubulk_name)

#copy uasset to workspace
def copy_uasset(folder, file, save_folder, clear=True):
    src_file = os.path.join(folder, file)
    TextureUasset(src_file) #check if the asset can parse

    #make or clear workspace
    save_folder = 'workspace/uasset'
    if clear and os.path.exists(save_folder):
        shutil.rmtree(save_folder)
        print('clear: {}'.format(save_folder))
    mkdir(save_folder)

    #copy files
    uasset_name, uexp_name, ubulk_name = get_all_file_path(src_file)
    new_file = os.path.join(save_folder, file)    
    new_uasset_name, new_uexp_name, new_ubulk_name = get_all_file_path(new_file)

    folder = os.path.dirname(new_file)
    if folder not in ['.', ''] and not os.path.exists(folder):
        mkdir(folder)
    shutil.copy(uasset_name, new_uasset_name)
    shutil.copy(uexp_name, new_uexp_name)
    print('copy: {} -> {}'.format(uasset_name, new_uasset_name))
    print('copy: {} -> {}'.format(uexp_name, new_uexp_name))
    if os.path.exists(ubulk_name):
        shutil.copy(ubulk_name, new_ubulk_name)
        print('copy: {} -> {}'.format(ubulk_name, new_ubulk_name))

#inject dds into the asset copied to workspace
def inject_dds(folder, file, save_folder, clear=True):
    uasset_folder = 'workspace/uasset'
    if not os.path.exists(uasset_folder):
        raise RuntimeError('Uasset Not Found.')

    #determine which file should be injected
    file_list = get_file_list_rec(uasset_folder)
    uexp_list=[]
    for f in file_list:
        if f[-4:]=='uexp':
            uexp_list.append(f)
    if len(uexp_list)==0:
        raise RuntimeError('Uasset Not Found.')
    elif len(uexp_list)==1:
        uasset_base=uexp_list[0]
    else:
        dds_base = os.path.splitext(os.path.basename(file))[0]
        uexp_base_list = [os.path.basename(file) for file in uexp_list]
        if dds_base+'.uexp' not in uexp_base_list:
            raise RuntimeError('The same name asset as dds not found. {}'.format(dds_base))
        id = uexp_base_list.index(dds_base+'.uexp')
        if id<0:
            raise RuntimeError('Uasset Not Found ({})'.format(os.path.join(uasset_folder, dds_base+'.uexp')))
        uasset_base=uexp_list[id]

    #read uasset
    uasset_file = os.path.join(uasset_folder, uasset_base)
    texture = TextureUasset(uasset_file)

    #read and inject dds
    src_file = os.path.join(folder, file)
    new_file = os.path.join(save_folder, uasset_base)
    dds = DDS.load(src_file)
    texture.inject_dds(dds)
    texture.save(new_file)

#export uasset as dds
def export_as_dds(folder, file, save_folder, clear=True):
    src_file = os.path.join(folder, file)
    new_file = os.path.join(save_folder, file)
    new_file=os.path.splitext(new_file)[0]+'.dds'

    texture = TextureUasset(src_file)
    dds = DDS.asset_to_DDS(texture)
    dds.save(new_file)

#remove mipmaps from uasset
def remove_mipmaps(folder, file, save_folder, clear=True):
    src_file = os.path.join(folder, file)
    new_file = os.path.join(save_folder, file)
    print(save_folder)
    print(file)
    print(new_file)
    texture = TextureUasset(src_file)
    texture.remove_mipmaps()
    texture.save(new_file)

mode_functions = {'valid': valid,
         'copy_uasset': copy_uasset,
         'inject': inject_dds,
         'remove_mipmaps': remove_mipmaps,
         'parse': parse,
         'export': export_as_dds,
         }

#main
if __name__=='__main__':
    args = get_args()
    file = args.file
    save_folder = args.save_folder
    mode = args.mode

    try:
        if mode not in mode_functions:
            raise RuntimeError('Unsupported mode. {}'.format(mode))
        func = mode_functions[mode]

        if os.path.isfile(file) and file[-3:]!='txt':
            #if input is a file
            folder = os.path.dirname(file)
            file = os.path.basename(file)
            func(folder, file, save_folder)

        else:
            if os.path.isfile(file):
                #if input file is txt (file list)
                folder, file_list = get_file_list_from_txt(file)
                func= [copy_uasset, inject_dds]
                inject=0
                for file in file_list:
                    func[inject](folder, file, save_folder)
                    inject = not inject
            else:
                #if input is a folder
                folder = file
                clear=True
                folder, file_list = get_file_list_from_folder(file)
                for file in file_list:
                    if file[-4:]=='uexp' or file[-3:] in ['dds', 'DDS']:
                        func(folder, file, save_folder, clear=clear)
                        clear=False

    except Exception as e:
        print('Error: {}'.format(e))
        raise RuntimeError(e)
    print('Success!')



src/uasset.py

from io_util import *

class UassetHeader:
    HEAD = b'\xC1\x83\x2A\x9E'

    def __init__(self, f):
        head=f.read(4)
        check(head, UassetHeader.HEAD, f, 'NOT a uasset file.')
        self.version=-read_int32(f)-1
        check(self.version, 6, f, 'Unsupported version. (version {})'.format(self.version))
        read_null_array(f, 4, 'Parse Failed.')
        self.file_size=read_uint32(f)
        none=read_str(f)
        check(none, 'None', f, 'Parse Failed.')
        self.unk_ary=read_uint8_array(f, len=4)

        self.name_num = read_uint32(f)
        self.name_offset = read_uint32(f)
        check(self.name_offset, 193, f, 'Parse Failed.')
        read_null_array(f, 2, 'Parse Failed.')

        self.export_num = read_uint32(f)
        self.export_offset = read_uint32(f)
        self.import_num = read_uint32(f)
        self.import_offset = read_uint32(f)
        self.unk1=f.read(4)
        read_null_array(f, 4, 'Parse Failed.')

        self.guid_hash=f.read(16)

        self.unk2=f.read(8)

        name_num=read_uint32(f)
        check(name_num, self.name_num, f, 'Parse Failed.')
        read_null_array(f, 9, 'Parse Failed.')
        self.unk3=f.read(4)
        read_null(f, 'Parse Failed.')
        self.padding_offset = read_uint32(f)
        self.file_length=read_uint32(f)
        read_null_array(f, 3, 'Parse Failed.')
        self.unk4=f.read(4)
        self.file_data_offset = read_uint32(f)

    def read(f):
        return UassetHeader(f)
    
    def write(f, header):
        f.write(UassetHeader.HEAD)
        write_int32(f, -(header.version+1))
        write_null_array(f, 4)
        write_uint32(f, header.file_size)
        write_str(f, 'None')
        write_uint8_array(f, header.unk_ary)
        write_uint32(f, header.name_num)
        write_uint32(f, header.name_offset)
        write_null_array(f, 2)
        write_uint32(f, header.export_num)
        write_uint32(f, header.export_offset)
        write_uint32(f, header.import_num)
        write_uint32(f, header.import_offset)
        f.write(header.unk1)
        write_null_array(f, 4)
        f.write(header.guid_hash)
        f.write(header.unk2)
        write_uint32(f, header.name_num)
        write_null_array(f, 9)
        f.write(header.unk3)
        write_null(f)
        write_uint32(f, header.padding_offset)
        write_uint32(f, header.file_length)
        write_null_array(f, 3)
        f.write(header.unk4)
        write_uint32(f, header.file_data_offset)


    def print(self):
        print('Header info')
        print('  version: {}'.format(self.version))
        print('  file size: {}'.format(self.file_size))
        print('  number of names: {}'.format(self.name_num))
        print('  name directory offset: {}'.format(self.name_offset))
        print('  number of exports: {}'.format(self.export_num))
        print('  export directory offset: {}'.format(self.export_offset))
        print('  number of imports: {}'.format(self.import_num))
        print('  import directory offset: {}'.format(self.import_offset))
        print('  guid hash: {}'.format(self.guid_hash))
        print('  padding offset: {}'.format(self.padding_offset))
        print('  file length (uasset+uexp-4): {}'.format(self.file_length))
        print('  file data offset: {}'.format(self.file_data_offset))

class UassetImport: #28 bytes
    def __init__(self, f):
        self.bin1=f.read(8)
        self.class_id=read_uint32(f)
        self.bin2=f.read(8)
        self.name_id=read_uint32(f)
        self.bin3=f.read(4)
        self.material=False

    def read(f):
        return UassetImport(f)
    
    def write(f, import_):
        f.write(import_.bin1)
        write_uint32(f, import_.class_id)
        f.write(import_.bin2)
        write_uint32(f, import_.name_id)
        f.write(import_.bin3)

    def name_imports(imports, name_list):
        for import_ in imports:
            import_.name=name_list[import_.name_id]
            import_.class_name=name_list[import_.class_id]

    def print(self, padding=2):
        pad=' '*padding
        print(pad+self.name)
        print(pad+'  class: '+self.class_name)

class UassetExport: #104 bytes
    KNOWN_EXPORTS=['EndEmissiveColorUserData', 'SQEX_BonamikAssetUserData', 'SQEX_KineDriver_AssetUserData', 'SkelMeshBoneAttributeRedirectorUserData', 'BodySetup']
    IGNORE=[True, True, True, True, True]
    #'BodySetup'
    def __init__(self, f):
        self.bin1=f.read(16)
        self.name_id=read_uint32(f)
        self.bin2=f.read(8)
        self.size=read_uint32(f)
        read_null(f)
        self.offset=read_uint32(f)
        self.bin3=f.read(64)

    def read(f):
        return UassetExport(f)
    
    def write(f, export):
        f.write(export.bin1)
        write_uint32(f, export.name_id)
        f.write(export.bin2)
        write_uint32(f, export.size)
        write_null(f)
        write_uint32(f, export.offset)
        f.write(export.bin3)

    def update(self, size, offset):
        self.size=size
        self.offset=offset

    def name_exports(exports, name_list, file_name):
        for export in exports:
            name=name_list[export.name_id]

            if name in UassetExport.KNOWN_EXPORTS:
                export.id=UassetExport.KNOWN_EXPORTS.index(name)
                export.ignore=UassetExport.IGNORE[export.id]
            elif name in file_name:
                export.id=-1
                export.ignore=False
            else:
                raise RuntimeError('Unsupported assets. ({})'.format(name))

            export.name=name

    def read_uexp(self, f):
        self.bin=f.read(self.size)

    def write_uexp(self, f):
        f.write(self.bin)

    def print(self, padding=2):
        pad=' '*padding
        print(pad+self.name)
        print(pad+'  size: {}'.format(self.size))
        print(pad+'  offset: {}'.format(self.offset))


class Uasset:

    def __init__(self, uasset_file, verbose=False):
        if uasset_file[-7:]!='.uasset':
            raise RuntimeError('Not .uasset. ({})'.format(uasset_file))

        if verbose:
            print('Loading '+uasset_file+'...')

        self.file=os.path.basename(uasset_file)[:-7]
        f=open(uasset_file, 'rb')
        self.size=get_size(f)
        self.header=UassetHeader.read(f)
        self.bin1 = f.read(self.header.name_offset-193)
        if verbose:
            print('size: {}'.format(self.size))
            self.header.print()
            print('Name list')
        
        self.name_list = []
        self.flag_list = []
        for i in range(self.header.name_num):
            name = read_str(f)
            flag = f.read(4)
            if verbose:
                print('  {}: {}'.format(i, name))
            self.name_list.append(name)
            self.flag_list.append(flag)
        offset=f.tell()
        self.bin2=f.read(self.header.import_offset-offset)

        self.imports=read_array(f, UassetImport.read, len=self.header.import_num)
        UassetImport.name_imports(self.imports, self.name_list)
        if verbose:
            print('Import')
            for import_ in self.imports:
                import_.print()

        offset=f.tell()
        self.bin3=f.read(self.header.export_offset-offset)
        self.exports=read_array(f, UassetExport.read, len=self.header.export_num)
        UassetExport.name_exports(self.exports, self.name_list, self.file)

        if verbose:
            print('Export')
            for export in self.exports:
                export.print()

        self.bin4=f.read()
        f.close()
    
    def save(self, file, uexp_size):
        self.header.file_length=uexp_size+self.size-4
        print('save :' + file)
        with open(file, 'wb') as f:
            UassetHeader.write(f, self.header)
            f.write(self.bin1)
            for name, flag in zip(self.name_list, self.flag_list):
                write_str(f, name)
                f.write(flag)

            f.write(self.bin2)
            write_array(f, self.imports, UassetImport.write)                
            f.write(self.bin3)
            write_array(f, self.exports, UassetExport.write)
            f.write(self.bin4)

src/io_util.py
import os

def mkdir(dir):
    os.makedirs(dir, exist_ok=True)

def get_size(file):
    pos=file.tell()
    file.seek(0,2)
    size=file.tell()
    file.seek(pos)
    return size

def check(actual, expected, f=None, msg=''):
    if actual!=expected:
        if f is not None:
            print('offset: {}'.format(f.tell()))
        print('actual: {}'.format(actual))
        print('expected: {}'.format(expected))
        raise RuntimeError(msg)

def read_uint32(file):
    bin=file.read(4)
    return int.from_bytes(bin, "little")

def read_uint16(file):
    bin=file.read(2)
    return int.from_bytes(bin, "little")

def read_uint8(file):
    bin=file.read(1)
    return int(bin[0])

def read_int32(file):
    bin=file.read(4)
    return int.from_bytes(bin, "little", signed=True)

def read_array(file, read_func, len=None):
    if len is None:
        len = read_uint32(file)
    ary=[read_func(file) for i in range(len)]
    return ary

def read_uint32_array(file, len=None):
    return read_array(file, read_uint32, len=len)

def read_uint16_array(file, len=None):
    return read_array(file, read_uint16, len=len)

def read_uint8_array(file, len=None):
    return read_array(file, read_uint8, len=len)

def read_int32_array(file, len=None):
    return read_array(file, read_int32, len=len)

def read_str(file):
    num = read_uint32(file)
    if num==0:
        return None
    string = file.read(num-1).decode()
    file.seek(1,1)
    return string

def read_const_uint32(f, n, msg='Unexpected Value!'):
    const = read_uint32(f)
    check(const, n, f, msg)

def read_null(f, msg='Not NULL!'):
    read_const_uint32(f, 0, msg)

def read_null_array(f, len, msg='Not NULL!'):
    null=read_uint32_array(f, len=len)
    check(null, [0]*len, f, msg)

def write_uint32(file, n):
    bin = n.to_bytes(4, byteorder="little")
    file.write(bin)

def write_uint16(file, n):
    bin = n.to_bytes(2, byteorder="little")
    file.write(bin)

def write_uint8(file, n):
    bin = n.to_bytes(1, byteorder="little")
    file.write(bin)

def write_int32(file, n):
    bin = n.to_bytes(4, byteorder="little", signed=True)
    file.write(bin)

def write_array(file, ary, write_func, with_length=False):
    if with_length:
        write_uint32(file, len(ary))
    for a in ary:
        write_func(file, a)

def write_uint32_array(file, ary, with_length=False):
    write_array(file, ary, write_uint32, with_length=with_length)

def write_uint16_array(file, ary, with_length=False):
    write_array(file, ary, write_uint16, with_length=with_length)

def write_uint8_array(file, ary, with_length=False):
    write_array(file, ary, write_uint8, with_length=with_length)

def write_int32_array(file, ary, with_length=False):
    write_array(file, ary, write_int32, with_length=with_length)

def write_str(file, s):
    num = len(s)+1
    write_uint32(file, num)
    str_byte = s.encode()
    file.write(str_byte + b'\x00')

def write_null(f):
    write_uint32(f, 0)

def write_null_array(f, len):
    write_uint32_array(f, [0]*len)

def compare(file1,file2):
    f1=open(file1, 'rb')
    f2=open(file2, 'rb')
    print('Comparing {} and {}...'.format(file1, file2))

    f1_size=get_size(f1)
    f2_size=get_size(f2)
    
    size=min(f1_size, f2_size)
    i=0
    f1_bin=f1.read()
    f2_bin=f2.read()
    f1.close()
    f2.close()

    i=-1
    for b1, b2 in zip(f1_bin, f2_bin):
        i+=1
        if b1!=b2:
            break
  
    if i==size-1:
        print('Same data!')
    else:
        raise RuntimeError('Not same :{}'.format(i))

    if f1_size!=f2_size:
        raise RuntimeError('Not same size. ({}, {})'.format(f1_size, f2_size))

src/file_list.py
import os

'''
format

1st line: folder path
2nd line: uexp
3rd line: dds
4th line: uexp
5th line: dds
    :      :

'''

'''
example

if file_list.txt is

foo/bar/
uexp/a.uexp
dds/d.dds
test/ppp.uexp
b/ttt.dds

then DDS tool will inject

foo/bar/dds/d.dds into foo/bar/uexp/a.uexp
foo/bar/b/ttt.dds into foo/bar/test/ppp.uexp
'''

def remove_quotes(l):
    l=l[:-1]
    if l in ['', '"']:
        return ''
    if l[0]=='"':
        l=l[1:]
    if l[-1]=='"':
        l=l[:-1]
    return l

def get_file_list_from_txt(file):
    with open(file, 'r') as f:
        lines = f.readlines()
    
    lines = [remove_quotes(l) for l in lines]
    lines = [l for l in lines if l!='']
    folder = lines[0]
    file_list = lines[1:]

    directory, folder = get_base_folder(folder)
    file_list = [os.path.join(folder, file) for file in file_list]
    return directory, file_list

def get_base_folder(p):
    folder = os.path.basename(p)
    directory = os.path.dirname(p)
    if folder=="":
        folder = os.path.dirname(directory)
        directory = os.path.basename(directory)  
    if directory==".":
        directory=""
    return directory, folder

def get_file_list_from_folder(folder):
    file_list = sorted(os.listdir(folder))
    directory, folder = get_base_folder(folder)
    file_list = [os.path.join(folder, file) for file in file_list]
    return directory, file_list

def get_file_list_rec(folder):
    file_list=[]
    for file in sorted(os.listdir(folder)):
        file_path = os.path.join(folder, file)
        if os.path.isdir(file_path):
            file_list+=[os.path.join(file, f) for f in get_file_list_rec(file_path)]
        else:
            file_list.append(file)
    return file_list



src/dds.py
import os, argparse
from io_util import *
from texture_asset import BYTE_PER_PIXEL

DDS_FORMAT = {
    'DXT1/BC1': ['DXT1', 71, 72],     #DXGI_FORMAT_BC1_TYPELESS
    'DXT5/BC3': ['DXT5', 77, 78],     #DXGI_FORMAT_BC3_TYPELESS	
    'BC4/ATI1': [80, 'ATI1', 'BC4U'], #DXGI_FORMAT_BC4_UNORM
    'BC5/ATI2': [83, 'ATI2', 'BC5U'], #DXGI_FORMAT_BC5_UNORM
    'BC6H(unsigned)': [95],           #DXGI_FORMAT_BC6H_UF16
    'BC6H(signed)': [96],             #DXGI_FORMAT_BC6H_SF16
    'BC7': [98, 99],                  #DXGI_FORMAT_BC7_TYPELESS
    'FloatRGBA': [10],                #DXGI_FORMAT_R16G16B16A16_FLOAT
    'B8G8R8A8(sRGB)': [91]            #DXGI_FORMAT_B8G8R8A8_UNORM_SRGB
}

def get_dds_format(form):
    for k in DDS_FORMAT:
        if form in DDS_FORMAT[k]:
            return k
    raise RuntimeError('Unsupported DDS format. ({})'.format(form))

class DDSHeader:
    MAGIC = b'\x44\x44\x53\x20'

    def __init__(self, width, height, mipmap_num, format_name):
        self.width = width
        self.height = height
        self.mipmap_num = mipmap_num
        self.format_name = format_name
        self.byte_per_pixel = BYTE_PER_PIXEL[self.format_name]

    #read header
    def read(f):
        head = f.read(4)             #Magic=='DDS '
        check(head, DDSHeader.MAGIC, msg='Not DDS.')
        read_const_uint32(f, 124)    #Size==124
        f.seek(4,1)                  #Flags
        height = read_uint32(f)      #Height
        width = read_uint32(f)       #Width
        f.seek(8,1)                  #PitchOrLinearSize, Depth
        mipmap_num = read_uint32(f)  #MipMapCount
        mipmap_num += mipmap_num==0
        f.seek(44, 1)                #Reserved1[11]
        read_const_uint32(f, 32)     #PfSize==32
        f.seek(4, 1)                 #PfFlags==4
        fourCC=f.read(4).decode()    #FourCC
        f.seek(40, 1)                #BitCount, BitMask
                                     #Caps, Caps2, ReservedCaps[2], Reserved2
        
        #DXT10 header
        if fourCC=='DX10':
            dxgi_format=read_uint32(f)      #dxgiFormat
            l = read_uint32_array(f, len=3) #resourceDimension==3
            check(l, [3,0,1])               #miscFlag==0
                                            #arraySize==1
            f.seek(4, 1)                    #miscFlag2
        else:
            dxgi_format=fourCC
        
        format_name = get_dds_format(dxgi_format)
        return DDSHeader(width, height, mipmap_num, format_name)

    #write header
    def write(f, header):
        mipmap_num=header.mipmap_num
        dxgi_format = DDS_FORMAT[header.format_name][0]
        if type(dxgi_format)==type(''):
            fourCC=dxgi_format
        else:
            fourCC='DX10'

        f.write(DDSHeader.MAGIC)
        write_uint32(f, 124) #Size==124

        #flags
        write_uint8(f, 7)
        write_uint8(f, 16)
        write_uint8(f, 8+2*(mipmap_num>1))
        write_uint8(f, 0)

        write_uint32(f, header.height)
        write_uint32(f, header.width)

        #PitchOrLinearSize
        write_uint32(f, int(header.width*header.height*header.byte_per_pixel))

        write_uint32(f, 1) #Depth
        write_uint32(f, mipmap_num)

        #Reserved1[11]
        write_null_array(f, 9)
        f.write('FF7R'.encode())
        write_null(f)

        write_uint32(f, 32) #PfSize==32
        write_uint32(f, 4) #PfFlags
        f.write(fourCC.encode()) #fourCC
        write_null_array(f, 5) #BitCount, BitMask

        #caps
        write_uint8(f, (mipmap_num>1)*8)
        write_uint8(f, 16)
        write_uint8(f, (mipmap_num>1)*64)
        write_uint8(f, 0)

        write_null_array(f ,4) #caps2, reservedCaps, reserved2

        #write dxt10 header
        if fourCC=='DX10':
            write_uint32(f, dxgi_format)
            write_uint32_array(f, [3,0,1])
            write_uint32(f, 0)

    def print(self):
        print('  height: {}'.format(self.height))
        print('  width: {}'.format(self.width))
        print('  format: {}'.format(self.format_name))
        print('  mipmap num: {}'.format(self.mipmap_num))
        #print('  byte per pixel: {}'.format(self.byte_per_pixel))

class DDS:
    def __init__(self, header, mipmap_data, mipmap_size):
        self.header = header
        self.mipmap_data = mipmap_data
        self.mipmap_size = mipmap_size

    #load dds file
    def load(file, verbose=False):
        if file[-3:] not in ['dds', 'DDS']:
            raise RuntimeError('Not DDS.')
        print('load: ' + file)
        with open(file, 'rb') as f:
            #read header
            header = DDSHeader.read(f)

            mipmap_num = header.mipmap_num
            byte_per_pixel = header.byte_per_pixel
            height = header.height
            width = header.width
            mipmap_data = []
            mipmap_size = []

            #read mipmaps
            for i in range(mipmap_num):

                #mipmap sizes are multiples of 4
                if height%4!=0:
                    height+=4-height%4
                if width%4!=0:
                    width+=4-width%4
                
                #read mipmap data
                size = height*width*byte_per_pixel
                if size!=int(size):
                    raise RuntimeError('The size of mipmap data is not int. This is unexpected.')
                data = f.read(int(size))
                
                #print mipmap info
                if verbose:
                    print('  Mipmap {}'.format(i))
                    print('    size (w, h): ({}, {})'.format(width, height))

                #store mipmap data
                mipmap_data.append(data)
                mipmap_size.append([int(width), int(height)])
                height = height//2
                width = width//2

            header.print()
            check(f.tell(), get_size(f), msg='Parse Failed. This is unexpected.')

        return DDS(header, mipmap_data, mipmap_size)
            
    #texture asset to dds
    def asset_to_DDS(asset):
        #make dds header
        header = DDSHeader(0, 0, 0, asset.format_name)

        mipmap_data=[]
        mipmap_size=[]

        #get ubulk mipmaps
        if asset.has_ubulk:
            for d, meta in zip(asset.ubulk_map_data, asset.ubulk_map_meta):
                mipmap_data.append(d)
                mipmap_size.append([meta.width, meta.height])

        #get uexp mipmaps
        for d, meta in zip(asset.uexp_map_data, asset.uexp_map_meta):
            mipmap_data.append(d)
            mipmap_size.append([meta.width, meta.height])

        #update header
        header.width, header.height = asset.get_max_size()
        header.mipmap_num=len(mipmap_data)

        return DDS(header, mipmap_data, mipmap_size)

    #save as dds
    def save(self, file):
        folder = os.path.dirname(file)
        if folder not in ['.', ''] and not os.path.exists(folder):
            mkdir(folder)

        with open(file, 'wb') as f:
            #write header
            DDSHeader.write(f, self.header)

            #write mipmap data
            for d in self.mipmap_data:
                f.write(d)
